server/engine/src/main/java/org/eclipse/lsp/cobol/implicitDialects/cics/utility/CICSReceiveOptionsCheckUtility.java

```diff
@@ -98,32 +84,25 @@ public <E extends ParserRuleContext> void checkOptions(E ctx) {
   }
 
   private void checkGroupOne(CICSParser.Cics_receive_group_oneContext ctx) {
-    checkHasMandatoryOptions(ctx.cics_length_flength(), ctx, "LENGTH or FLENGTH");
-    if (!ctx.ASIS().isEmpty() || !ctx.BUFFER().isEmpty()) {
-      checkHasIllegalOptions(ctx.LEAVEKB(), "LEAVEKB");
-    }
-  }
-
-  private void checkGroupTwo(CICSParser.Cics_receive_group_twoContext ctx) {
-    checkHasMandatoryOptions(ctx.cics_length_flength(), ctx, "LENGTH or FLENGTH");
-    checkHasMandatoryOptions(ctx.cics_into_set(), ctx, "INTO or SET");
-  }
-
-  private void checkGroupThree(CICSParser.Cics_receive_group_threeContext ctx) {
-    checkHasMandatoryOptions(ctx.cics_length_flength(), ctx, "LENGTH or FLENGTH");
+    checkHasMutuallyExclusiveOptions("INTO or SET", ctx.INTO(), ctx.SET());
+    if (!ctx.SET().isEmpty())
+      checkHasExactlyOneOption("LENGTH or FLENGTH", ctx, ctx.cics_length_flength());
+    checkHasMutuallyExclusiveOptions("MAXLENGTH or MAXFLENGTH", ctx.cics_maxlength());
   }
 
   private void checkPartn(CICSParser.Cics_receive_partnContext ctx) {
-    checkHasMandatoryOptions(ctx.cics_into_set(), ctx, "INTO or SET");
+    checkHasMandatoryOptions(ctx.PARTN(), ctx, "PARTN");
+    checkHasExactlyOneOption("INTO or SET", ctx, ctx.cics_into_set());
     checkHasMandatoryOptions(ctx.LENGTH(), ctx, "LENGTH");
   }
 
   private void checkMap(CICSParser.Cics_receive_mapContext ctx) {
-    if (ctx.FROM().isEmpty()) checkHasIllegalOptions(ctx.LENGTH(), "LENGTH");
-    if (ctx.TERMINAL().isEmpty()) checkHasIllegalOptions(ctx.INPARTN(), "INPARTN");
+    if (ctx.FROM().isEmpty()) checkHasIllegalOptions(ctx.LENGTH(), "LENGTH without FROM");
+    checkHasMutuallyExclusiveOptions("INTO or SET", ctx.cics_into_set());
```

Comment:
Something similar here, I'd expect `checkIntoSet(ctx.cics_into_set())` and then in it, length check + `checkHasMutuallyExclusiveOptions("...", ctx.get(0).INTO(), ctx.get(0).SET())`